mlflow/projects/__init__.py

```diff
@@ -233,6 +237,8 @@ def _fetch_git_repo(uri, version, dst_dir, git_username, git_password):
     origin.fetch()
     if version is not None:
```

Comment:
Thanks for the PR! One question, doesn't this branch of the existing code already handle checking out a specific branch of the repo? As in, passing the name of a branch as the version argument via `mlflow run -v <branch-name>` should work already.

Reply:
Thanks @smurching. 

Yes, it works since the `repo.git.checkout` produces the same for version and branch. However, there are two reasons that I think it deserves as a separate option. First, the description of `-v` or `--version` option is for commit version. If it is used for specifying a branch as well, could we change the name and description? Secondly, when the project generates the `source_version`, it calls `_get_git_commit` to retrieve the commit. And then the source_version is rendered to the UI. This means that the `<branch-name>` will lose even it has been specified through the `-v` option.

Reply:
Got it, yeah I think updating the docstring for the `--version` option to explicitly state that it can be the name of a branch makes a ton of sense (we should do so in the CLI & the `projects.run` API). Sorry for the confusion on that!

Your point about wanting to save the branch name is also a good one. I think one way to save branch names is via a tag. In the future we'll likely extend run search to allow searching over tags, so you'd be able to do things like filter runs by git branch name, etc.  IMO we should continue to store the git version as a commit hash, since commit hashes can be used to recover the exact source code used for the run (whereas branch names don't correspond to a specific commit).

If we want to go the tag route, we could:
* Create a new internal tag for git branches in [mlflow_tags.py](https://github.com/mlflow/mlflow/blob/master/mlflow/utils/mlflow_tags.py) (e.g. `MLFLOW_GIT_BRANCH_NAME = "mlflow.gitBranchName"`)
* Add an `_is_valid_branch_name(project_dir, branch_name)` helper to mlflow/projects/__init__.py, and use it to verify if `version` is a branch name. The implementation could use logic like `repo = Repo(project_dir, search_parent_directories=True); repo.git.rev_parse("--verify", "refs/heads/%s" % branch_name)` (approach based on [this SO post](https://stackoverflow.com/q/5167957))
* If `version` is a branch name as per our helper function, set the branch-name tag [here](https://github.com/mlflow/mlflow/blob/HEAD/mlflow/projects/__init__.py#L54).

Let me know if that makes sense & if you'd be interested in taking a stab at it, & thanks again for bringing this up!

Reply:
@smurching thanks for the detailed elaboration and I like the idea of internal `branch` tag, which sounds better than my original thought of introducing `source_branch`. This does satisfy my need to specify source by branch and be able to query the runs with branch name.

I will work on this based on your suggestion above. Thanks again.